ENIGMAsystem/SHELL/Universal_System/spritestruct.cpp

```diff
@@ -17,687 +17,460 @@
 
-#define get_current_instance() \
-    ((enigma::object_graphics*) enigma::instance_event_iterator->inst)
+ResourceVec<Sprite> sprites("sprite");
 
-bool get_sprite(enigma::sprite* &spr, int id)
-{
-#ifdef DEBUG_MODE
-    if (id < -1 || size_t(id) > enigma::sprite_idmax || !enigma::spritestructarray[id]) {
-        show_error("Cannot access sprite with id " + toString(id), false);
-        return false;
-    }
-#endif
-    spr = enigma::spritestructarray[id];
-    return true;
+void Sprite::freeTextures() {
+  for (unsigned i = 0; i < subimages.size(); ++i) enigma::graphics_delete_texture(subimages[i].textureID);
 }
 
-bool get_sprite_mtx(enigma::sprite* &spr, int id)
-{
-    bool rtn = get_sprite(spr, id);
-    if (rtn) {
-        // TODO: Lock a lock by reference and allow it to be timely destructed and released
-    }
-    return rtn;
-}
+size_t Sprite::subcount() { return subimages.size(); }
 
-namespace enigma {
-  sprite** spritestructarray;
-  extern size_t sprite_idmax;
-  sprite::sprite() {}
-  sprite::sprite(int x) {}
+SubImage::SubImage(const SubImage &copy) {
+  textureID = graphics_duplicate_texture(copy.textureID);
+  x = copy.x;
+  y = copy.y;
+  w = copy.w;
+  h = copy.h;
+  for (void *c : copy.collisionData) collisionData.push_back(c);
 }
 
-namespace enigma_user
-{
+inline void lowerLeftToTransparent(unsigned char *pxdata, unsigned w, unsigned h) {
+  int idx = (h - 1) * w * 4;
+  int color[3];
+  color[0] = pxdata[idx];
+  color[1] = pxdata[idx + 1];
+  color[2] = pxdata[idx + 2];
+  for (unsigned i = 0; i < w * h * 4; i += 4) {
+    idx = (h - 1) * w * 4;
+    int color2[3];
+    color2[0] = pxdata[idx];
+    color2[1] = pxdata[idx + 1];
+    color2[2] = pxdata[idx + 2];
 
-int sprite_add(string filename, int imgnumb, bool precise, bool transparent, bool smooth, bool preload, int x_offset, int y_offset, bool mipmap)
-{
-    enigma::spritestructarray_reallocate();
-    enigma::sprite *spr = enigma::spritestructarray[enigma::sprite_idmax] = new enigma::sprite();
-    enigma::sprite_add_to_index(spr, filename, imgnumb, precise, transparent, smooth, x_offset, y_offset, mipmap);
-    return enigma::sprite_idmax++;
+    if (color == color2) pxdata[idx + 3] = 0;
+  }
 }
 
-int sprite_add(string filename, int imgnumb, bool transparent, bool smooth, int x_offset, int y_offset, bool mipmap)
-{
-    return sprite_add(filename, imgnumb, false, transparent, smooth, true, x_offset, y_offset, mipmap);
-}
+inline std::array<bbox_rect_t, 2> calculateBBox(unsigned w, unsigned h, int xoffset, int yoffset) {
+  // FIXME: Calculate and assign correct bbox values.
+  std::array<bbox_rect_t, 2> b;
+
+  b[0].bottom = h;
+  b[0].left = 0;
+  b[0].top = 0;
+  b[0].right = w;
+
+  b[1].bottom = b[0].bottom - yoffset;
+  b[1].left = b[0].left - xoffset;
+  b[1].top = b[0].top - yoffset;
+  b[1].right = b[0].right - xoffset;
+
+  return b;
+}
+
+inline void addTexture(Sprite *spr, unsigned char *pxdata, unsigned fullWidth, unsigned fullHeight,
```

Comment:
You're computing `fullWidth`/`fullHeight` now; why are you also passing them as read-only?